plugin/documents.py

```diff
@@ -621,31 +630,25 @@ def _do_highlights_async(self) -> None:
             session.send_request_async(request, self._on_highlights)
 
     def _on_highlights(self, response: Optional[List]) -> None:
-        if not response:
-            self._clear_highlight_regions()
-            return
-        kind2regions = {}  # type: Dict[int, List[sublime.Region]]
-        for kind in range(1, 4):
-            kind2regions[kind] = []
+        if not isinstance(response, list):
+            response = []
+        kind2regions = {}  # type: Dict[Tuple[int, bool], List[sublime.Region]]
         for highlight in response:
             r = range_to_region(Range.from_lsp(highlight["range"]), self.view)
             kind = highlight.get("kind", DocumentHighlightKind.Text)
-            if kind in kind2regions:
-                kind2regions[kind].append(r)
-            else:
-                debug("unknown DocumentHighlightKind", kind)
+            kind2regions.setdefault((kind, len(self.view.split_by_newlines(r)) > 1), []).append(r)
 
         def render_highlights_on_main_thread() -> None:
             self._clear_highlight_regions()
-            flags = userprefs().document_highlight_style_to_add_regions_flags()
-            for kind, regions in kind2regions.items():
+            flags_multi, flags_single = userprefs().document_highlight_style_to_add_regions_flags()
+            for tup, regions in kind2regions.items():
+                kind, multiline = tup
+                key = self._highlights_key(kind, multiline)
                 if regions:
-                    scope = _kind2scope[kind]
-                    self.view.add_regions(
-                        "lsp_highlight_{}".format(_kind2name[kind]),
-                        regions,
-                        scope=scope,
-                        flags=flags)
+                    flags = flags_multi if multiline else flags_single
+                    self.view.add_regions(key, regions, scope=_kind2scope[kind], flags=flags)
+                else:
+                    self.view.erase_regions(key)
```

Comment:
There is a `self._clear_highlight_regions()` call at the beginning. Is this really needed?